### ISSUE TITLE
[Regression] Perfomance drop after LLVM 16.0.0 update (#13534)

### ISSUE THREAD

**Issue by Ordinary205** [RPCS3/rpcs3#13638]
### Quick summary

After updating RPCS3, the performance gets decreased.

### Details

Before: 14845
![Test 2](https://user-images.githubusercontent.com/125112932/230906597-e16afd5c-45c2-4047-8306-e03244710f5b.PNG)

After: 14854
![Test 4](https://user-images.githubusercontent.com/125112932/230906639-943db26a-834a-41fd-ae9b-064a58b51145.PNG)

### Build with regression

v0.0.27-14854-6fbca1ac

### Attach two log files

14845: 
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/11190622/RPCS3.log.gz)

14854:
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/11190623/RPCS3.log.gz)

### System configuration

AMD Ryzen 5900X 12-Core Processor | 24 Threads | 15.89 GiB RAM | RTX 3080 driver 531.41.0.0 | Windows 10 Pro 22H2

**Comment by AniLeo** [comment]
Not valid as range has more than one build, might be related to the LLVM update but we need the exact build

Also, needs retest with #13629

**Comment by solarmystic** [comment]
Definitely seems in line with the minor performance regressions i noted in the LLVM16 PR, but please test with 14846 to confirm that the LLVM16 PR is the responsible one @Ordinary205 https://github.com/RPCS3/rpcs3-binaries-win/releases/download/build-fb88e1c1c933a792f96600206ed03a07cd7c3661/rpcs3-v0.0.27-14846-fb88e1c1_win64.7z

**Comment by Ordinary205** [comment]
> Definitely seems in line with the minor performance regressions i noted in the LLVM16 PR, but please test with 14846 to confirm that the LLVM16 PR is the responsible one @Ordinary205 https://github.com/RPCS3/rpcs3-binaries-win/releases/download/build-fb88e1c1c933a792f96600206ed03a07cd7c3661/rpcs3-v0.0.27-14846-fb88e1c1_win64.7z

I didn't notice that the 14846 build was causing the regression. I thought it was the recent update but nevermind, I'll test the 14846.

**Comment by Ordinary205** [comment]
Heres the correct results:

Before - 14845:
![Test 6](https://user-images.githubusercontent.com/125112932/230914093-4567fc50-fd0d-477d-bd4e-96fcee0569d5.PNG)

After - 14846:
![Test 5](https://user-images.githubusercontent.com/125112932/230912512-505877b0-04d7-4f1a-aff9-744405c85867.PNG)

14845: [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/11190852/RPCS3.log.gz)
14846: [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/11190791/RPCS3.log.gz)

**Comment by Ordinary205** [comment]
I can confirm that the 14846 is a regression.

**Comment by solarmystic** [comment]
Thanks for the verification that the LLVM16 PR is the responsible one.

**Comment by Darkhost1999** [comment]
Is the performance restored in https://github.com/RPCS3/rpcs3/pull/13629

**Comment by solarmystic** [comment]
> @Darkhost1999 I did some performance tests using 14845 and 14855.

Looking at your Infamous 2 screenshot results, performance is improved now and back to normal?

14845 - 33.88 FPS
14846 (the LLVM16 build) - 29.24 FPS
14855 - 35.99 FPS

So I'd say the Infamous 2 performance regression is fixed right?

**Comment by Ordinary205** [comment]
I'd say the performance on Infamous 2 is almost fixed.

**Comment by Asinin3** [comment]
DeS isn't a good test, the framerate counter doesn't work properly in that game due to how it works.
Your GoW 3 and Infamous test are showing basically the same avg fps despite major fluctuations and the usual run-to-run variance. The regression to performance looks like it was fixed by 14855 due to neko properly reverting a upstream LLVM commit that we identified as lowering performance (he reverted it in the 16.0.1 PR)

Can't see any reason to leave this open.

**Comment by Nekotekina** [comment]
I think some issue still remains. I didn't revert upstream commit (it's not technically possible), I only did a workaround

**Comment by Asinin3** [comment]
Ok re-opened. But infamous fps difference on avg is 0.1 and GoW3 was 1.5 (but only after a random drop). Both games fluctuate a lot, i think RDR would be a better test or TLoU.. something that doesn't fluctuate a lot.

**Comment by Ordinary205** [comment]
Should I test Red dead redemption?

**Comment by Asinin3** [comment]
Yes

**Comment by Ordinary205** [comment]
So I tested Red Dead Redemption on 14845 and 14855.

14845:
![Test 15 (2)](https://user-images.githubusercontent.com/125112932/231838254-52edad31-35d1-4391-aba8-372a89a98989.jpg)
![Test 16 (2)](https://user-images.githubusercontent.com/125112932/231838270-9e3b695d-3ad2-4079-a1e5-eeaf120bbd5a.jpg)

14855:
![Test 17 (2)](https://user-images.githubusercontent.com/125112932/231838291-83542d0d-45db-45bb-8d21-600f090d8e96.jpg)
![Test 18 (2)](https://user-images.githubusercontent.com/125112932/231838307-da55b17b-47c4-4eaf-9c51-85e9ccf856e4.jpg)

**Comment by Asinin3** [comment]
Performance will need to be tested again in 14876. Expect performance to go down, because neko had to revert something to fix a regression

**Comment by Ordinary205** [comment]
I might have to test 14876 on The Last of Us aswell.

**Comment by Ordinary205** [comment]
Heres the results for Red Dead Redemption and The Last of Us.

Red Dead Redemption:
14845:
![Test 15 (2)](https://user-images.githubusercontent.com/125112932/232224617-458a826e-4e96-4d11-b0aa-fe85e9342418.jpg)
![Test 16 (2)](https://user-images.githubusercontent.com/125112932/232224597-799e8270-31cc-4c6c-8593-45914ae388a6.jpg)

14876:
![rpcs3 exe 2023 04 15 - 16 16 51 02 - frame at 1m42s (2)](https://user-images.githubusercontent.com/125112932/232224304-99f58c41-d4e1-4fe9-a80a-c6abd34b7c85.jpg)
![rpcs3 exe 2023 04 15 - 16 16 51 02 - frame at 3m26s (2)](https://user-images.githubusercontent.com/125112932/232224310-085930c3-4791-4672-89b9-2e7b0cc6631e.jpg)


The Last of Us:
14845:
![rpcs3 exe 2023 04 15 - 16 32 13 04 - frame at 1m19s (2)](https://user-images.githubusercontent.com/125112932/232224344-0dbe3337-d212-42e5-a31e-140745964fc8.jpg)

14876:
![rpcs3 exe 2023 04 15 - 16 26 26 03 - frame at 0m59s (2)](https://user-images.githubusercontent.com/125112932/232224340-305fc0e3-5978-4e7a-ba5b-8465b32fc226.jpg)


Red Dead Redemption:
14845: [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/11239519/RPCS3.log.gz)
14876: [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/11239515/RPCS3.log.gz)

The Last of Us:
14845: [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/11239518/RPCS3.log.gz)
14876: [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/11239514/RPCS3.log.gz)

**Comment by solarmystic** [comment]
Borrowing this Issue report to note that God of War 3 and God of War Ascension have also regressed in performance since #13670 got merged to master with 0.0.27-14876.

GOWA

14875 - 65/60/55 FPS
![GOW3 875](https://user-images.githubusercontent.com/4345150/232229773-b1be20fc-ab5b-4e08-8950-9359a60ef633.PNG)

14876 - 61/55/51 FPS
![GOW3 876](https://user-images.githubusercontent.com/4345150/232229778-a05b6e75-3af4-4313-ac28-9044cac31f14.PNG)


GOW3

14875 -129/95/70 FPS
![GOWA 875](https://user-images.githubusercontent.com/4345150/232229787-7b757bdf-bec6-49c0-b188-bd389abc0fe9.PNG)


14876 - 122/97/64 FPS
![GOWA 876](https://user-images.githubusercontent.com/4345150/232229798-05d3f761-15af-4918-9881-f803fe917f7a.PNG)

**Comment by Ordinary205** [comment]
Good thing that I'm not the only one who is experiencing a performance loss.

**Comment by Asinin3** [comment]
Test with https://github.com/RPCS3/rpcs3/pull/13672 Windows build just finished a few seconds ago

**Comment by Ordinary205** [comment]
@Asinin3 Sorry if this is a dumb question, but how do I test this PR build?

**Comment by Blackbird88** [comment]
> @Asinin3 Sorry if this is a dumb question, but how do I test this PR build?

Download the artifact [here](https://cirrus-ci.com/task/5930585584041984)

**Comment by Ordinary205** [comment]
> > @Asinin3 Sorry if this is a dumb question, but how do I test this PR build?
> 
> Download the artifact [here](https://cirrus-ci.com/task/5930585584041984)

Thanks

**Comment by Asinin3** [comment]
Don't link people to artifacts. That commit is already old, and we want people testing the latest commits on PRs.
To download a PR build just go to the PR: https://github.com/RPCS3/rpcs3/pull/13672 > go to checks tab at top or bottom of page, then go to Cirrus Windows > View More Details on Cirrus > Artifact > download the artifact for the emulator, not the SHA file.

**Comment by solarmystic** [comment]
Another update to this issue report - after #13672 got merged to master in v0.0.27-14881, performance drops have been fixed for GOW3 and GOW Ascension, they are back to within margin of error compared to 14845.

The performance regression in Persona 5 that exists since LLVM16 got merged to master still remains.

14845 (last pre-LLVM16 master) - 97/72/63 FPS

![P5 14845](https://user-images.githubusercontent.com/4345150/232801001-7057ea3d-1d6d-400d-8686-c4ff6d632886.PNG)

14881(current master with #13672)

![P5 14881](https://user-images.githubusercontent.com/4345150/232801170-28d3a8ac-b75b-4a16-9623-8ab7b727c25a.PNG)

**Comment by Ordinary205** [comment]
Sadly it still has performance regression on Persona 5, even after 14898 was released.

**Comment by cipherxof** [comment]
MGS4 is also suffering from performance loss after the LLVM 16 upgrade.

<details>
  <summary>Pre LLVM 16</summary>
  
![3](https://github.com/RPCS3/rpcs3/assets/5994581/aae23507-0e5b-4edf-bdd9-1b77d6349178)
![2](https://github.com/RPCS3/rpcs3/assets/5994581/0f199a0e-bfff-4588-82fd-12d5edb80ec2)

</details>

<details>
  <summary>LLVM 16</summary>
  
![3](https://github.com/RPCS3/rpcs3/assets/5994581/ea50f935-de15-4162-b2be-629bc8d30b1c)
![2](https://github.com/RPCS3/rpcs3/assets/5994581/2e2725a4-93b7-4400-b15f-5fe316b257bc)
</details>

The master build is about the same.

[RPCS3-13534.log.gz](https://github.com/RPCS3/rpcs3/files/11638495/RPCS3-13534.log.gz)
[RPCS3-13613.log.gz](https://github.com/RPCS3/rpcs3/files/11638496/RPCS3-13613.log.gz)

**Comment by cipherxof** [comment]
It turns out that at least for Linux performance is restored when using the `rocketlake` LLVM profile instead of `icelake-client`. I'll have to retest this on Windows when I get a chance.

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/11858098/RPCS3.log.gz)

![Screenshot from 2023-06-24 13-31-38](https://github.com/RPCS3/rpcs3/assets/5994581/a81bb932-e166-4bd9-b481-6354f3cdc25c)

![Screenshot from 2023-06-24 13-33-02](https://github.com/RPCS3/rpcs3/assets/5994581/e2499af3-c413-4ea2-89d9-f34566f811a2)

**Comment by cipherxof** [comment]
This is still an issue on Windows for my machine. Initially I thought it was related to AVX-512, so I decided to test without it and alas performance is still better on LLVM 15.

![image](https://github.com/RPCS3/rpcs3/assets/5994581/bfbc33aa-217e-4cc9-8e8d-6ce92bc52d08)

**Comment by cipherxof** [comment]
It turns out the performance loss isn't due to LLVM 16 at all, but one specific change that was introduced in that same PR. The following change restores performance back to normal (+10% gains).

![image](https://github.com/RPCS3/rpcs3/assets/5994581/4ad6fbcd-a1cf-40ff-9901-e249944c0d87)

[RPCS3-tweaked.log.gz](https://github.com/RPCS3/rpcs3/files/12222443/RPCS3-tweaked.log.gz)

My local build used in the log also reverted another PR which resulted in further performance loss (https://github.com/RPCS3/rpcs3/issues/14386). With both of these changes performance is now better than it was 14845.
